Implement indexed access to config line fields.

rvandegrift opened the pull request:
If x is a BaseCfgLine representing 'ip access-group block-stuff in',
x[0] will return 'ip', x[2] will return 'block-stuff' and x[2:4] will
return 'block-stuff in'.

mpenning commented:
Thank you for offering a contribution.  Please help me understand why this is useful and why cdp_abc should override __getitem__()

rvandegrift commented:
On 2015-01-16 13:47, Mike Pennington wrote:
> Thank you for offering a contribution. Please help me understand why
> this is useful and why cdp_abc should override GETITEM()

I frequently use ciscoconfparse to extract info from a config and audit 
or report on it.  To produce human readable output, it's useful to 
extract the names of things from the config.  Currently, that means 
using .text.split() each time.  Overloading __getitem__ is a cosmetic 
improvement for that use case.

Here's an example.  Suppose I'm pulling in info about interfaces and 
ACLs for a report:

interfaces = []
c = CiscoConfParse('router23')
print 'Report on %s' % c.find_object('hostname ')[0].text.split()[1]
for i in c.find_objects(r'interface '):
     interface = {}
     interface['name'] = i.text.split()[1]
     interface['prefixes'] = [IPv4Network(x.text.split()[2:4] for x in 
i.re_search_children(r'ip address ')]

     aclin = i.re_search_children(r'ip access-group .* in')
     if aclin:
         interfaces['acl'] = c.find_objects(r'ip access-list.* %s' % 
aclin.text.split()[2]
     interfaces.append(interface)
...

With the patch, the ".text.split()" can be dropped.  In my mind, the 
improvement shows up in the list comprehension.

Thanks,
Ross

mpenning commented:
Quite honestly, you're trying to do this the hard way with `split()` on string fields...

1.  Parse configs with `CiscoConfParse('/path/to/config', factory=True)`; I am slowly documenting everything you can do with the `factory` option (see the [`IOSIntfLine()` docs](http://www.pennington.net/py/ciscoconfparse/api_IOSIntfLine.html) for examples, but the bottom line is that ciscoconfparse tries to do a lot of the work for you when use parse with the `factory` option.

2.  Assuming you have assigned an interface line object to `i` and parsed with `factory`, `interface['name'] = i.text.split()[1]` is very simply written as `interface['name'] = i.name`

I had to try to decipher what you were trying to do in your example above, because several things were broken in the example, but this is what I came up with...


    from ciscoconfparse import CiscoConfParse

    interfaces = []
    c = CiscoConfParse('sample_01.ios', factory=True)
    print 'Report on %s' % c.find_objects(r'^hostname ')[0].text.split()[1]
    for i in c.find_objects(r'^interface '):
         interface = {}
         interface['name'] = i.name
        prefixes = []
         interface['prefixes'] = [(j.re_match('address\s(\S+)'), j.re_match('address\s\S+\s(\S+)')) for j in i.re_search_children(r'^\s*ip\saddr')]

         aclin = i.re_match_iter_typed(r'ip access-group (\S+) in', result_type=str)
         if aclin:
             acl_lines = c.find_all_children(r'ip access-list (extended )*%s' % aclin)
             interface['acl'] = acl_lines
         interfaces.append(interface)


At this point, I don't see much value in overridding `ccp_abc.__getitem__()`; it makes the code more complicated and magical than I want to maintain.
